packages/drawer/src/ux-modal-service.ts

```diff
@@ -0,0 +1,277 @@
+import { EventAggregator } from 'aurelia-event-aggregator';
+import { UxDrawerTheme } from './ux-drawer-theme';
+import { UxDrawer } from './ux-drawer';
+import { DrawerPosition, DrawerKeybord, DefaultDrawerConfiguration } from './drawer-configuration';
+import { inject, TemplatingEngine, Controller } from 'aurelia-framework';
+
+export interface ModalServiceOptions {
+  viewModel?: any;
+  view?: any;
+  host?: Element| 'body' | string;
+  position?: DrawerPosition;
+  moveToBodyTag?: boolean;
+  theme?: UxDrawerTheme;
+  model?: any;
+  overlayDismiss?: boolean;
+  keyboard?: DrawerKeybord;
+  restoreFocus?: (lastActiveElement: HTMLElement) => void;
+}
+
+export interface ModalServiceResult {
+  wasCancelled: boolean;
+  output: any;
+}
+
+export interface ModalServiceDrawer {
+  whenClosed: () => Promise<ModalServiceResult>
+}
+
+interface ModalBindingContext {
+  composeViewModel?: {
+    viewModel?: any;
+    view?: any;
+    model?: any;
+    currentViewModel?: {
+      canDeactivate?: (result: any) => any;
+      deactivate?: (result: any) => any;
+      detached?: (result: any) => any;
+    };
+  };
+  theme?: UxDrawerTheme;
+  keyboard?: DrawerKeybord;
+  restoreFocus?: (lastActiveElement: HTMLElement) => void;
+  host?: HTMLElement;
+  dismiss?: () => void;
+  ok?: (event: Event) => void;
+}
+
+interface ModalLayer {
+  bindingContext?: ModalBindingContext;
+  drawer: UxDrawer;
+}
+
+@inject(TemplatingEngine, EventAggregator, DefaultDrawerConfiguration)
+export class ModalService {
+
+  public startingZIndex: number = 200;
+  public modalLayers: Array<ModalLayer> = [];
+  public drawerIndex: number = 0;
+
+  constructor(
+    private templatingEngine: TemplatingEngine, 
+    private eventAggregator: EventAggregator,
+    private defaultConfig: DefaultDrawerConfiguration) {
+
+  }
+
+  public addLayer(drawer: UxDrawer, bindingContext: ModalBindingContext) {
+    this.modalLayers.push({
```

Comment:
Passing the tip around:
```suggestion
    const layerCount = this.modalLayers.push({
      bindingContext: bindingContext,
      drawer: drawer
    });
    if (layerCount === 1) {
      this.setKeyListener();
    }
```